processing/src/main/java/org/apache/druid/query/expression/NestedDataExpressions.java

```diff
@@ -442,38 +442,76 @@ public String name()
+        if (castTo == null) {
+          throw new IAE("Invalid output type: [%s]", args.get(2).getLiteralValue());
         }
-
-        @Override
-        public ExprEval eval(ObjectBinding bindings)
+        class CastJsonValueExpr extends ExprMacroTable.BaseScalarMacroFunctionExpr
         {
-          ExprEval input = args.get(0).eval(bindings);
-          return ExprEval.bestEffortOf(
-              NestedPathFinder.findLiteral(unwrap(input), parts)
-          );
-        }
+          public CastJsonValueExpr(List<Expr> args)
+          {
+            super(name(), args);
+          }
 
-        @Override
-        public Expr visit(Shuttle shuttle)
-        {
-          List<Expr> newArgs = args.stream().map(x -> x.visit(shuttle)).collect(Collectors.toList());
-          return shuttle.visit(new JsonValueExpr(newArgs));
-        }
+          @Override
+          public ExprEval eval(ObjectBinding bindings)
+          {
+            ExprEval input = args.get(0).eval(bindings);
+            return ExprEval.bestEffortOf(
+                NestedPathFinder.findLiteral(unwrap(input), parts)
+            ).castTo(castTo);
+          }
 
-        @Nullable
-        @Override
-        public ExpressionType getOutputType(InputBindingInspector inspector)
+          @Override
+          public Expr visit(Shuttle shuttle)
+          {
+            List<Expr> newArgs = args.stream().map(x -> x.visit(shuttle)).collect(Collectors.toList());
+            return shuttle.visit(new CastJsonValueExpr(newArgs));
+          }
+
+          @Nullable
+          @Override
+          public ExpressionType getOutputType(InputBindingInspector inspector)
+          {
+            return castTo;
+          }
+        }
+        return new CastJsonValueExpr(args);
+      } else {
+        class JsonValueExpr extends ExprMacroTable.BaseScalarMacroFunctionExpr
         {
-          // we cannot infer the output type (well we could say it is 'STRING' right now because is all we support...
-          return null;
+          public JsonValueExpr(List<Expr> args)
+          {
+            super(name(), args);
+          }
+
+          @Override
+          public ExprEval eval(ObjectBinding bindings)
+          {
+            ExprEval input = args.get(0).eval(bindings);
+            return ExprEval.bestEffortOf(
+                NestedPathFinder.findLiteral(unwrap(input), parts)
+            );
+          }
+
+          @Override
+          public Expr visit(Shuttle shuttle)
+          {
+            List<Expr> newArgs = args.stream().map(x -> x.visit(shuttle)).collect(Collectors.toList());
+            return shuttle.visit(new JsonValueExpr(newArgs));
+          }
+
+          @Nullable
+          @Override
+          public ExpressionType getOutputType(InputBindingInspector inspector)
+          {
+            // we cannot infer the output type (well we could say it is 'STRING' right now because is all we support...
```

Comment:
Then why not return STRING?

Reply:
oh this is a stale comment, from a way earlier prototype when i did in fact only store string columns lol. Will remove since it is no longer true.

Reply:
👍